### PR TITLE
solution

### PR DESCRIPTION
**Pull request by vladyslav78292**
[DEMO LINK](https://vladyslav78292.github.io/react_todo-app-loading-todos/)

### CODE REVIEW

File: src/App.tsx [hunk 1]
```diff
@@ -1,156 +1,72 @@
...
+    getTodosById();
+  }, []);
+
   if (!USER_ID) {
     return <UserWarning />;
   }
 
+  const visibleTodos = getVisibleTodos(todos, filter);
+  const isThereCompleted = todos.some(todo => todo.completed);
+  const activeTodosAmount = todos.filter(todo => !todo.completed).length;
```
Comment (Bogdan-Kutasevich) [inline src/App.tsx:43]: there better to use useCallback and useMemo to optimize your components render.

File: src/App.tsx [hunk 2]
```diff
@@ -1,156 +1,72 @@
...
-        Unable to add a todo
-        <br />
-        Unable to delete a todo
-        <br />
-        Unable to update a todo
-      </div>
+      {isError && (
+        <ErrorMessage
+          errorType={errorType}
+          setError={setError}
```
Comment (Bogdan-Kutasevich) [inline src/App.tsx:67]: Same, better to use wrapper

**Review by Bogdan-Kutasevich — APPROVED** [review]
Good!)
Some comments below.

File: src/App.tsx [hunk 3]
```diff
@@ -1,156 +1,72 @@
...
+import { ErrorMessage } from './components/ErrorMessage/ErrorMessage';
 
-const USER_ID = 0;
+const USER_ID = 6390;
 
 export const App: React.FC = () => {
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [isError, setError] = useState(false);
+  const [filter, setFilter] = useState<Filter>(Filter.ALL);
+  const [errorType, setErrorType] = useState('');
```
Comment (alexandra-protyanova) [inline src/App.tsx:19]: let use enum here as well

File: src/App.tsx [hunk 4]
```diff
@@ -1,156 +1,72 @@
...
+
+  useEffect(() => {
+    const getTodosById = async () => {
+      try {
+        setError(false);
```
Comment (alexandra-protyanova) [inline src/App.tsx:24]: useless, it is false by default

File: src/App.tsx (hunk 1 above)
Comment (alexandra-protyanova) [reply src/App.tsx:43]: useMemo +

File: src/App.tsx [hunk 5]
```diff
@@ -1,156 +1,72 @@
...
+    };
+
+    getTodosById();
+  }, []);
+
   if (!USER_ID) {
     return <UserWarning />;
   }
 
+  const visibleTodos = getVisibleTodos(todos, filter);
```
Comment (alexandra-protyanova) [inline src/App.tsx:41]: useMemo here

File: src/components/ErrorMessage/ErrorMessage.tsx [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+}) => {
+  useEffect(() => {
+    window.setTimeout(() => {
+      setError(false);
+    }, 3000);
+  }, []);
+
+  return (
+    <div className="notification is-danger is-light has-text-weight-normal">
+      {/* eslint-disable-next-line jsx-a11y/control-has-associated-label */}
```
Comment (alexandra-protyanova) [inline src/components/ErrorMessage/ErrorMessage.tsx:20]: fix linter

File: src/components/ErrorMessage/ErrorMessage.tsx [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+        onClick={() => {
+          setError(false);
+        }}
+      />
+      {errorType}
+      {/* Unable to add a todo
+      <br />
+      Unable to delete a todo
+      <br />
+      Unable to update a todo */}
```
Comment (alexandra-protyanova) [inline src/components/ErrorMessage/ErrorMessage.tsx:33]: move to enum

File: src/components/Footer/Footer.tsx [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+    <nav className="filter">
+      <a
+        href="#/"
+        className={cn('filter__link', {
+          selected: filter === Filter.ALL,
+        })}
+        onClick={() => setFilter(Filter.ALL)}
+      >
+        All
+      </a>
```
Comment (alexandra-protyanova) [inline src/components/Footer/Footer.tsx:33]: you duplicate all `a`s, just map enum

File: src/components/Header/Header.tsx [hunk 9]
```diff
@@ -0,0 +1,21 @@
+import React from 'react';
+
+export const Header: React.FC = () => {
+  return (
```
Comment (alexandra-protyanova) [inline src/components/Header/Header.tsx:4]: redundant return

File: src/components/TodoInfo/TodoInfo.tsx [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+
+type Props = {
+  todo: Todo,
+};
+
+export const TodoInfo: React.FC<Props> = React.memo(({
+  todo,
+}) => {
+  const { title, completed } = todo;
+  const [isEdited] = useState(false);
```
Comment (alexandra-protyanova) [inline src/components/TodoInfo/TodoInfo.tsx:13]: will it be updated?

File: src/utils/getVisibleTodos.ts [hunk 11]
```diff
@@ -0,0 +1,17 @@
+import { Todo } from '../types/Todo';
+import { Filter } from '../types/filter';
+
+export const getVisibleTodos = (todos: Todo[], filterBy: Filter): Todo[] => {
+  return todos.filter(todo => {
```
Comment (alexandra-protyanova) [inline src/utils/getVisibleTodos.ts:5]: redundant return

File: src/utils/getVisibleTodos.ts [hunk 12]
```diff
@@ -0,0 +1,17 @@
...
+    switch (filterBy) {
+      case Filter.ALL:
+        return todo;
+      case Filter.COMPLETED:
+        return todo.completed;
+      case Filter.ACTIVE:
+        return !todo.completed;
+      default:
```
Comment (alexandra-protyanova) [inline src/utils/getVisibleTodos.ts:13]: you will never have unexpected result with typescript, move `All` to default case

**Review by alexandra-protyanova — CHANGES_REQUESTED** [review]

File: src/components/Header/Header.tsx (hunk 9 above)
Comment (vladyslav78292) [reply src/components/Header/Header.tsx:4]: i need it for the future

**Pull request closed without merging**